Print diagnostics on ProductResolver's loadTag setIntervalFor and IOVProxy's fetchSequence

JanChyczynski opened the pull request:
#### PR description:

Printing more diagnostics to help with investigating https://github.com/cms-sw/cmssw/issues/45555
Continuation of https://github.com/cms-sw/cmssw/pull/46395

Allows to define a list of records using `recordsToDebug` parameter of `CondDBESSource` for which additional debug information is printed:
* When loading a tag or when setting IOV interval (in ProductResolverWrapperBase)
* When fetching IOV sequence (in IOVProxy)

#### PR validation:

Tested with:

```
#!/bin/bash -ex

hltGetConfiguration /dev/CMSSW_15_0_0/GRun \
   --globaltag 150X_dataRun3_HLT_v1 \
   --data \
   --unprescale \
   --output minimal \
   --max-events 10 \
   --eras Run3_2024 --l1-emulator uGT --l1 L1Menu_Collisions2024_v1_3_0_xml \
   --input /store/data/Run2024I/EphemeralHLTPhysics0/RAW/v1/000/386/593/00000/91a08676-199e-404c-9957-f72772ef1354.root \
   > hlt.py


cat <<@EOF >> hlt.py
process.options.wantSummary = True
process.options.numberOfThreads = 1
process.options.numberOfStreams = 0
@EOF
```

Now the hlt.py produced by above code was edited to provide the list of records to debug:
add `recordsToDebug = cms.untracked.vstring("BeamSpotOnlineHLTObjectsRcd")`
to the parameters of `PoolDBESSource`:
```
process.GlobalTag = cms.ESSource( "PoolDBESSource",
    ...
    pfnPrefix = cms.untracked.string( "" ),
    recordsToDebug = cms.untracked.vstring("BeamSpotOnlineHLTObjectsRcd")
)
```

Now it can be run by:
```
cmsRun hlt.py &> hlt.log
```


Additionally tested on CMSSW_14_0_11 with a scenario directly from #45555:
```
#!/bin/bash -ex

# originally on CMSSW_14_0_11_MULTIARCHS

hltGetConfiguration run:383631 \
--globaltag 140X_dataRun3_HLT_v3 \
--data \
--no-prescale \
--no-output \
--max-events -1 \
--input '/store/group/tsg/FOG/error_stream_root/run383631/run383631_ls0444_index000423_fu-c2b14-43-01_pid675389.root,/store/group/tsg/FOG/error_stream_root/run383631/run383631_ls0445_index000025_fu-c2b14-43-01_pid675389.root,/store/group/tsg/FOG/error_stream_root/run383631/run383631_ls0666_index000047_fu-c2b14-43-01_pid675067.root,/store/group/tsg/FOG/error_stream_root/run383631/run383631_ls0445_index000002_fu-c2b14-43-01_pid675389.root,/store/group/tsg/FOG/error_stream_root/run383631/run383631_ls0666_index000027_fu-c2b14-43-01_pid675067.root,/store/group/tsg/FOG/error_stream_root/run383631/run383631_ls0666_index000065_fu-c2b14-43-01_pid675067.root' > hlt.py

cat <<@EOF >> hlt.py
process.options.wantSummary = True
process.options.numberOfThreads = 1
process.options.numberOfStreams = 0
@EOF
```


#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

15_0: #47619

cmsbuild commented:
Pull request #47171 was updated.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"052b7b6b67c36aa3533f14f2442fb94af24bcdba":{"files":["CondCore/CondDB/src/IOVProxy.cc"],"squashed":false,"time":1737649211},"42df8617fe77589a40434ce9a12afff4c490ef54":{"files":["CondCore/CondDB/BuildFile.xml","CondCore/CondDB/src/IOVProxy.cc"],"squashed":false,"time":1737541802},"4cd810d8838db1c37100c3a797597e134fd10cdb":{"files":["CondCore/CondDB/src/IOVProxy.cc"],"squashed":false,"time":1737647350},"654eb03a07d95a74f02e7fe074a8d447b18daf68":{"files":["CondCore/CondDB/BuildFile.xml","CondCore/CondDB/src/IOVProxy.cc","CondFormats/BeamSpotObjects/BuildFile.xml"],"squashed":false,"time":1737625577},"773ab5ab07c51481ee51b7cc0650ae3b430c9b5d":{"files":["CondCore/CondDB/src/IOVProxy.cc"],"squashed":false,"time":1737107884},"a69bc28053c7379f298d3f65921afa212bcd706d":{"files":["CondCore/CondDB/src/IOVProxy.cc"],"squashed":false,"time":1737627530},"ee76c5ae15ce6e9c9c79fd1a0fc1fc74530222c9":{"files":["CondCore/CondDB/src/IOVProxy.cc"],"squashed":false,"time":1737378496}},"emoji":{},"last_seen_sha":"052b7b6b67c36aa3533f14f2442fb94af24bcdba","signatures":{}} -->

JanChyczynski commented:
@Dr15Jones @mmusich Please take a look

FYI @PonIlya

cmsbuild commented:
Pull request #47171 was updated.

JanChyczynski commented on CondCore/CondDB/src/IOVProxy.cc:
```diff
@@ -292,6 +300,135 @@ namespace cond {
+          beamSpot.covariance().Det2(det);
+          if (vertexFromBSInvalid || inversionFail) {
+            std::ostringstream s;  //tmp, just for better code formatting
+            s << "\n--------------------- BeamSpot Diagnostics (hash: " << hash << ") ----------------\n"
+              << "Fetched invalid beamspot obj of hash " << hash << "\n"
+              << "Tag: " << iovProxyData->tagInfo.name << "\n"
+              << "IOV: " << iov << "(" << cond::time::timeTypeName(iovProxyData->tagInfo.timeType) << ")\n"
+              << "Request info: "
+              << "' request interval [ " << lowerGroup << " , " << higherGroup << " ] "
+              << "new range [ " << iovProxyData->groupLowerIov << " , " << iovProxyData->groupHigherIov << " ] "
+              << "#entries " << iovProxyData->iovSequence.size() << "\n"
+              << "\n"
+              << "BeamSpotOnlineObjects:\n"
+              << *beamspotObj
+              << "\n"
+              //TODO probably no new info here, enough to preant only BeamSpotOnlineObjects
+              << "BeamSpot:\n"
+              << beamSpot << "\n"
+              << "BeamSpot (BS) rotatedCovariance3D:\n"
+              << beamSpot.rotatedCovariance3D() << "\n"
+              << "BS position: " << beamSpot.position() << "\n"
+              << "BS covariance Determinant: " << det << "\n"
+              << "Of matrix:  " << beamSpot.covariance() << "\n"
+              << "is VertexState(BS).error() invalid? (0=valid): " << vertexFromBSInvalid << "\n"
+              << "VertexState(BS).error().matrix(): \n"
+              << beamVertexState.error().matrix() << "\n"
+              << "err cxx() = " << beamVertexState.error().cxx() << "\n"
+              << "err cyy() = " << beamVertexState.error().cyy() << "\n"
+              << "err czz() = " << beamVertexState.error().czz() << "\n"
+              << "did BS.covariance() inversion failed? (0=ok, 1=failed): " << inversionFail << "\n"
+              << "BS.covariance().Inverse():\n"
+              << errorInverse << "\n"
+              << "\n-------------- end of BeamSpot Diagnostics (hash: " << hash << ") ----------------\n"
+              << std::endl;
+            edm::LogSystem("NewIOV") << s.str();
```
Let me know what diagnostics should I remove and if any info is missing.

JanChyczynski commented on CondCore/CondDB/src/IOVProxy.cc:
```diff
@@ -292,6 +300,135 @@ namespace cond {
+
+          double det;  //TODO do we check BS validity by det?
+          beamSpot.covariance().Det2(det);
+          if (vertexFromBSInvalid || inversionFail) {
```
Please notice that the diagnostics are handled differently from the way that implementation by @Dr15Jones worked. Info is printed only when an invalid BeamSpotOnlineObjects payload is fetched and there is no check whether the fetching happened for the first time or not. 

Should there be any info displayed even when fetch objects are valid?

JanChyczynski commented on CondCore/CondDB/src/IOVProxy.cc:
```diff
@@ -292,6 +300,135 @@ namespace cond {
+          std::unique_ptr<BeamSpotOnlineObjects> beamspotObj = session.fetchPayload<BeamSpotOnlineObjects>(hash);
+          reco::BeamSpot beamSpot = beamSpotFromBSOnlineObjects(beamspotObj);
+
+          //Check validity of BeamSpot by validating error of VertexState created from it
+          VertexState beamVertexState(beamSpot);
+          bool vertexFromBSInvalid = (beamVertexState.error().cxx() <= 0.) || (beamVertexState.error().cyy() <= 0.) ||
+                                     (beamVertexState.error().czz() <= 0.);
+
+          //Check validity of BeamSpot by confirming it can be inverted
+          int inversionFail = 2;  //the .Inverse(...) will set it to 0 or 1
+          reco::BeamSpot::CovarianceMatrix errorInverse = beamSpot.covariance().Inverse(inversionFail);
+
+          double det;  //TODO do we check BS validity by det?
+          beamSpot.covariance().Det2(det);
```
Is that a good way to check the validity of beamSpotObjects? Wouldn't that be too big of an overhead for HLT? What would you do differently?

cmsbuild commented:
Pull request #47171 was updated.

cmsbuild commented:
Pull request #47171 was updated.

Dr15Jones commented:
I personally think this is way too intrusive on the base code (i.e. making the PoolDBESSource directly dependent upon beam spot object code).  Plus fetching ALL the beam spot object each time the IOV sequence is fetched is going to be pulling lots of data potentially from the DB all the time.

Why not instead add a parameter to PoolDBESSource to turn on printing the IOV info (i.e. iov/hash) in the fetch request if the `payloadType` matches a certain name? Such a change would then potentially help other debugging jobs. 

For this case, when the problem happens you can check what the IOV info was, see if it matches what was expected (and or seen by other worker nodes). If the hash doesn't match expectations then that identifies the in the IOV sequence request. If it does match it means the request to fetch using the 'hash' went wrong.

JanChyczynski committed:
print diagnostics based on m_printDebug flag to IOVProxy and ProductResolverWrapperBase

Allow to debug all records by passing a *

cleanup IOVProxy ctors

lumiIdToLumiNr and lumiIdToRun: implement and use in IOVProxy

cmsbuild commented:
Pull request #47171 was updated.

JanChyczynski commented on CondCore/ESSources/plugins/CondDBESSource.h:
```diff
@@ -120,6 +120,8 @@ class CondDBESSource : public edm::eventsetup::ESProductResolverProvider, public
   typedef std::map<std::string, cond::GTEntry_t> TagCollection;
   // the collections of tag, record/label used in this ESSource
   TagCollection m_tagCollection;
+  std::vector<std::string> m_recordsToDebug;
```
I'm considering changing it to recordAndLabelsToDebug and eg. let the user pass a vector of strings of records concatenated with labels separated by "@" (eg. `"MyRecord1Rcd@Label1", "MyRecord1Rcd@Label2", "MyRecord2Rcd"`)

Do you think it would be worth implementing in this PR? Maybe you have an idea of how to do it more intuitively for the user?

JanChyczynski replied:
I think we can leave it as it is for now and if there's a need for more precise choice we can implement it then

JanChyczynski commented on CondCore/CondDB/src/IOVProxy.cc:
```diff
@@ -292,6 +293,23 @@ namespace cond {
+      for (const auto& [iov, hash] : iovProxyData.iovSequence) {
+        s << "(" << iov << ", " << hash << ") ";
+      }
+      edm::LogSystem("NewIOV") << s.str() << std::endl;
```
is LogSystem the correct log level?

makortel replied:
The `LogSystem` is intended to be used only within the framework. The main difference for other levels is that it can not be suppressed (i.e. it does not respect any limits or thresholds). It is mentioned in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLoggerGuidelines and https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLoggerSeverities (but you'd have to know what to look for, so those don't count towards "(well) documented").

Conceptually this message seems like information, rather than error or warning. INFO-level messages are suppressed by default, so they would need to be explicitly enabled in the MessageLogger configuration. Given the severity of the issue being investigated, from `core` side I'd be fine with `LogSystem` here (as an exceptional case). From the application perspective, is there need or desire to limit these messages, or hide them by default?

@Dr15Jones Do you have any further thoughts?

mmusich replied:
>  The main difference for other levels is that it can not be suppressed (i.e. it does not respect any limits or thresholds). 

we definitely do not want to not be able to suppress it in the HLT menu.

Dr15Jones replied:
> we definitely do not want to not be able to suppress it in the HLT menu.

The messages will not be sent if the configuration option is to False (which is the default). My understanding is the point of this is to find the bug in the HLT running so this needs to be enabled for the BeamSpot issue.

JanChyczynski replied:
@makortel I've organized the code as suggested.

Is there anything more to discuss about the log level or can we resolve this thread?

makortel replied:
> I've organized the code as suggested.

Thanks. The last `s << std::endl;` seems redundant though (doubly so, MessageLogger itself adds a newline to the end of every message, and the message itself seems to end in an explicit newline even without that line).

> Is there anything more to discuss about the log level or can we resolve this thread?

From `core` perspective we don't have anything to add to that discussion.

JanChyczynski reviewed (commented):
I've implemented the solution based on @Dr15Jones suggestions. Please take a look at my github comments to the code and lmk if you have suggestions. @Dr15Jones @mmusich

Dr15Jones commented on CondCore/CondDB/src/IOVProxy.cc:
```diff
@@ -313,8 +331,18 @@ namespace cond {
+        //                           << "IOVProxy m_printDebug: " << m_printDebug << std::endl;
+        if (printIOVSequenceDiagnostics(*m_data, lowerGroup, higherGroup)) {
+          //TODO make sure this is only in debug mode / doesn't get merged into the final version
+          std::exit(0);  //To avoid long exection of unrelated code after diagnostics got printed
```
I assume this is not supposed to be in the final PR?

cmsbuild commented:
Pull request #47171 was updated.

cmsbuild commented:
Pull request #47171 was updated.

cmsbuild commented:
Pull request #47171 was updated.

cmsbuild commented:
Pull request #47171 was updated.

cmsbuild commented:
This PR contains too many commits (6616 >= 240) and will not be processed.
Please ensure you have selected the correct target branch and consider squashing unnecessary commits.
The processing of this PR will resume once the commit count drops below the limit.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44134

cmsbuild commented:
A new Pull Request was created by  @JanChyczynski for master.

It involves the following packages:

- CondCore/CondDB (**db**)
- CondCore/ESSources (**db, alca**)


@atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please review it and eventually sign? Thanks.
@JanChyczynski, @PonIlya, @mmusich, @rsreds, @tocheng, @yuanchao this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

JanChyczynski commented:
From my side it's ready. LMK if you have any more suggestions
@Dr15Jones  @mmusich

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73d852/45051/summary.html
**COMMIT**: 4e5eb615cd5b7470bf45af550032936df9ddfcaa
**CMSSW**: CMSSW_15_1_X_2025-03-18-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47171/45051/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 223 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-18-1100+73d852/67496/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 5 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-18-1100+73d852/67496/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3927607
 - DQMHistoTests: [Total failures: 24](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-18-1100+73d852/67496/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3927563
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-18-1100+73d852/67496/triggerResults): no differences found

mmusich commented:
> You potentially [removed 223 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-18-1100+73d852/67496/validateJR/logRootQA-events.log) from the logs

is this the desired effect?

mmusich commented:
> . By default, no records are tracked so this results in less logs in the tests.

OK, so in order to get debug level information for specific records (after this PR is merged), we will need to edit the GlobalTag `ESSource` from within the HLT menu with something like

```python3
process.GlobalTag = cms.ESSource( "PoolDBESSource",
    ...

    recordsToDebug = cms.untracked.vstring("BeamSpotOnlineHLTObjectsRcd", "BeamSpotOnlineHLTObjectsRcd", "BeamSpotOnlineLegacyObjectsRcd")
)
```

@cms-sw/hlt-l2 @missirol FYI.

perrotta commented:
Shouldn't you add the new parameter also in CondCore/ESSources/python/CondDBESSource_cfi.py, then?
Default there can remain an empty string, but at least it will be explicitely listed in the python configuration

Martin-Grunewald commented:
This is not ConfDb frienly as the new parameters ```recordsToDebug``` is neither in a fillDescriptions method (maybe add a fillDescriptions method to the class) nor in any cfi file in ../python.

mmusich commented:
indeed adding a `fillDescriptions` to `CondCore/ESSources/plugins/CondDBESSource.cc` is the favored way (this doesn't prevent to add the parameters explicitly in `CondCore/ESSources/python/CondDBESSource_cfi.py, ` too.

mmusich commented:
>  indeed adding a fillDescriptions to `CondCore/ESSources/plugins/CondDBESSource.cc` is the favored way (this doesn't prevent to add the parameters explicitly in `CondCore/ESSources/python/CondDBESSource_cfi.py`, too.


https://github.com/cms-sw/cmssw/pull/47630 contains an attempt to do add a `fillDescriptions` to `CondDBESSource`

perrotta commented:
> > indeed adding a fillDescriptions to `CondCore/ESSources/plugins/CondDBESSource.cc` is the favored way (this doesn't prevent to add the parameters explicitly in `CondCore/ESSources/python/CondDBESSource_cfi.py`, too.
> 
> #47630 contains an attempt to do add a `fillDescriptions` to `CondDBESSource`

Thank you @mmusich !
I would be tempted to sign #47630 for alca and db, and then once that PR will get merged ask @JanChyczynski to rebase and update this one accordingly: or did you have a different plan in mind?

mmusich commented:
> or did you have a different plan in mind?

that works for me, if you are fine with https://github.com/cms-sw/cmssw/pull/47630

perrotta commented:
@JanChyczynski after the merge of #47630 this PR must be rebased, and adapted to the parameters being now initialized in the fillDescriptions method

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44211

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44211/cms-checkout-topic.log

mmusich commented on CondCore/ESSources/plugins/CondDBESSource.cc:
```diff
@@ -125,6 +125,8 @@ CondDBESSource::CondDBESSource(const edm::ParameterSet& iConfig)
       m_connection(),
       m_connectionString(""),
       m_frontierKey(""),
+      m_recordsToDebug(
```
this is still not in the `fillDescriptions`. Please add it.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44213

mmusich commented on CondCore/CondDB/interface/IOVProxy.h:
```diff
@@ -104,7 +104,8 @@ namespace cond {
 
-      //
+      // TODO This seems to violate The rule of three: why copy constructor and assignment is defined but not the descructor?
+      // Is this ctor even needed? How is it different from the implicitly defined one?
```
should this become an issue instead of a comment?

makortel replied:
It would be better to refer to the rule of five (i.e. including the move constructor and assignment as well). By quick look all of the four could just be defaulted (which also means they could be left to be implicitly declared by the constructor).

I think it would be good for this PR to do something for them, as in the present state of the PR the `m_printDebug` does not get forwarded through the copy operations.

JanChyczynski replied:
The default ctor and copy ctor that were implemented there were not different from what would be implicitly generated (until I added the `m_printDebug` member, thank you for a very good catch @makortel!)   

* I changed the explicit implementation of default ctor to `= default`
   * some doubts described later
* removed the explicit copy ctor, now it should be implicitly generated
* removed the explitic copy assign operator,  now it should be implicitly generated
* move assign and move ctor should be implicitly generated

Let me know if I made any mistakes.

The ctor from a session is commented with `// the only way to construct it from scratch...` - does it mean the default constructor should be removed?... I doubt it, but its a bit confusing

mmusich commented:
@cmsbuild, please test

mmusich commented:
I tried merging this PR locally on top of 

```bash
#!/bin/bash -ex

hltGetConfiguration /dev/CMSSW_15_0_0/GRun \
   --globaltag 150X_dataRun3_HLT_v1 \
   --data \
   --unprescale \
   --output minimal \
   --max-events 10 \
   --eras Run3_2024 --l1-emulator uGT --l1 L1Menu_Collisions2024_v1_3_0_xml \
   --input /store/data/Run2024I/EphemeralHLTPhysics0/RAW/v1/000/386/593/00000/91a08676-199e-404c-9957-f72772ef1354.root \
   > hlt.py


cat <<@EOF >> hlt.py
process.options.wantSummary = True
process.options.numberOfThreads = 1
process.options.numberOfStreams = 0
process.GlobalTag.recordsToDebug = cms.untracked.vstring("BeamSpotOnlineHLTObjectsRcd", "BeamSpotOnlineHLTObjectsRcd", "BeamSpotOnlineLegacyObjectsRcd")
@EOF

cmsRun hlt.py >& hlt.log 
```

and I see in the output:

```console
%MSG-s NewIOV:  PostBeginProcessBlock 24-Mar-2025 13:26:44 CET  PostBeginProcessBlock
Fetched new IOV for 'BeamSpotOnlineHLT'
payload type: BeamSpotOnlineObjects
request interval [ 1660399996895502 , 1660404291862534 ]
new range [ 1660399996895502 , 1660404291862533 ]
#entries 42
sequence (iov, hash): (1660399996895502, db9bcbf8e662319de9f370acae084c0581c00d9f) (1660399996895507, b1351d1934f319dd454c54f499bab3e29da2251c) (1660399996895512, afbc256a95d872c62b44d6f356c54ed0aed1f44d)
 (1660399996895517, fe8627be9fdadec718a60ea0a6b9d2ff83b71327) (1660399996895522, 360d30bd8b40c11b7903fef8644effa882a0e582) (1660399996895527, 0cbe48a3b8e9be2fd5bb95e49dc686872beca5ec) (1660399996895532, 5
e2e50a0934666a594a714ae1d6e23a8f729eab7) (1660399996895537, ddabfba5d71d80d6893dc16be4a715764bef3cf7) (1660399996895542, 549e1317967e1fcb7fa169408ce792be144c2278) (1660399996895547, e7595b2bda2d411f85660e
dca388eedf82d8c378) (1660399996895552, 59b13a90ed9fb1bb1be09022eb5308ea8f3b4c83) (1660399996895557, b5dc9523ad86a4e01112438c768911aa630cc039) (1660399996895562, 4d47183fc9d75afd875fd363b6e2bda1f74ebcfb) (
1660399996895567, 81cf0334bc32d06e42716bd2e85b6db46ba9f26d) (1660399996895572, b082409127a0d02583fec2c03842636875a5f244) (1660399996895577, f1dac7d263b0671ce4961cb04277ad745a2be22c) (1660399996895582, 9c8
4739ab1612659b97894f87acdf7e0e7c632e5) (1660399996895587, c21d6e7cb7eb7fcf15421e7d5ccf210a8ef4a773) (1660399996895592, 8fb575b172cff3ec688da42a6a6af2f6803e283f) (1660399996895597, bfad071465bd2a6848575443
bdbda267c5a5df3a) (1660399996895602, 5e3d1db76b0c88cdefade6719742c6240d1103d8) (1660399996895607, 269dbe432ef85fd79f725c2bc542151502eefb9f) (1660399996895612, 9c23c2c81bbb5e194e35abc3105defff60461b13) (16
60399996895617, 7736e40ac7ad88123011af1f4127fa38766ea80e) (1660399996895622, 30f95592e9f29d03e44f01c88b9fb2e34294dadb) (1660399996895627, 50714486364e76e01a68bc9043552486cce89e1f) (1660399996895632, 2ed1b
77adafeea4729fe781b63ac41d7c18320eb) (1660399996895637, 43a754c611011d2bc465268602fd641b1f720207) (1660399996895642, eaa76835e91b91df1a1a5a4d8464b947978972ae) (1660399996895647, cf75f62da5c53e33924b6f9c89
25105f30650010) (1660399996895652, ede51d1c7ad15e5196ee3c4a5d6498be4e2d0b83) (1660399996895657, 391be7d12c0fff51e75cd1ce6f1598aea42cd12f) (1660399996895661, f4e2887290efe0432bc110d219773ace244b1a6c) (1660
399996895667, ea70e527ca5c5c22cb59260ccd8c5abc081390e7) (1660399996895672, 07ffb657178af37e45b537eaf96dc2093a9f701a) (1660399996895677, f060509933a73cf973cabd295df0f85b0940fb96) (1660399996895682, b6ece50
8012a527e90618f035c4c5764a738219f) (1660399996895687, 1732b0cfd1de50997d774530e64f23a960301e4d) (1660399996895692, 549401263ee13f9ed4fcf6853ecd7592b8b3d7ba) (1660399996895697, 395a17b00bc4b837a3fd516ca15f
8386c5078497) (1660399996895702, c0e52241e7c806947a826b733de85a19057bc3d8) (1660399996895707, 3876d68dcc660c7df8d93acf87df7fad89a7af64) 
%MSG
```

some suggestions:
   * could the "since" of the iov be expressed in the unpacked way (run,ls), in addition to the packed one (for at least the fields `request interval`  and `new range`) ?
   * could the visual display of list of `sequence` be improved such that it displays in more readable way? maybe with line breaks?

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73d852/45171/summary.html
**COMMIT**: dc8fa6f2be4a47af970a64025ffd8e18ce6386ad
**CMSSW**: CMSSW_15_1_X_2025-03-24-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47171/45171/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 8 errors in the following unit tests:

<pre>---> test testDiamondLUTCalibration had ERRORS
---> test SiPixelGainCalibrationTests had ERRORS
---> test testCondToolsSiStripBuildersReaders had ERRORS
and more ...
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 223 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-24-1100+73d852/67638/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-24-1100+73d852/67638/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3909207
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-24-1100+73d852/67638/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3909187
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-24-1100+73d852/67638/triggerResults): no differences found

perrotta commented:
> I found 8 errors in the following unit tests:
> 
> ---> test testDiamondLUTCalibration had ERRORS
> ---> test SiPixelGainCalibrationTests had ERRORS
> ---> test testCondToolsSiStripBuildersReaders had ERRORS
> and more ...

Once the [suggestion here above](https://github.com/cms-sw/cmssw/pull/47171#issuecomment-2748002682) will have been implemented this PR will have to be tested together with #47675 (if not yet merged)

JanChyczynski commented on CondCore/ESSources/python/CondDBESSource_cfi.py:
```diff
@@ -20,5 +20,6 @@
     pfnPostfix       = '',
-    pfnPrefix        = '' ,
+    pfnPrefix        = '',
+    recordsToDebug   = [],
```
Should this remain empty or are there any records for which this logging should be printed by default?

mmusich replied:
as far as the application is only for HLT and only for the beam spot, I would keep this empty and let the individual clients customize as needed.

JanChyczynski commented:
> some suggestions:
> 
> * could the "since" of the iovs be expressed in the unpacked way (run,ls), in addition to the packed one (for at least the fields `request interval`  and `new range`) ?
> * could the visual display of list of `sequence (iov, hash)` be improved such that it displays in more readable way? maybe with line breaks?

@mmusich Good point! Now the logs look eg. like this:
```
Fetched new IOV for 'BeamSpotOnlineHLT'
payload type: BeamSpotOnlineObjects
request interval [ 1660399996895502 , 1660404291862534 ]
new range [ 1660399996895502 , 1660404291862533 ]
request interval (run, LS): [ (386592, 270) , (386593, 6) ]
new range (run, LS): [ (386592, 270) , (386593, 5) ]
#entries 42
sequence [iov (run, LS), hash]:
1660399996895502 (386592, 270), db9bcbf8e662319de9f370acae084c0581c00d9f,
1660399996895507 (386592, 275), b1351d1934f319dd454c54f499bab3e29da2251c,
1660399996895512 (386592, 280), afbc256a95d872c62b44d6f356c54ed0aed1f44d,
```

I've pushed a new version but I still have to address Markotel's comment that:
>`m_printDebug` does not get forwarded through the copy operations.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44217

cmsbuild commented:
Pull request #47171 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

makortel commented on CondCore/CondDB/src/IOVProxy.cc:
```diff
@@ -293,8 +293,38 @@ namespace cond {
         throwException("The transaction is not active.", ctx);
     }
 
+    bool printIOVSequenceDiagnostics(const IOVProxyData& iovProxyData,
```
This function seems to return `true` unconditionally, and the caller seems to ignore the return value. How about not returning anything?
```suggestion
    void printIOVSequenceDiagnostics(const IOVProxyData& iovProxyData,
```

mmusich commented:
@JanChyczynski @cms-sw/alca-l2 what's the status of this PR? 
We'd like to have the backport https://github.com/cms-sw/cmssw/pull/47619 in production for the first 900GeV beams.

JanChyczynski commented:
> @JanChyczynski @cms-sw/alca-l2 what's the status of this PR? We'd like to have the backport #47619 in production for the first 900GeV beams.

Sorry for slowing this down - I had to focus on more pressing issues last week. I'm working on the constructors of `IOVProxy` and some refactoring in `printIOVSequenceDiagnostics`. I'll push the changes later today.

JanChyczynski commented:
Addressed the comments but some refactoring in `printIOVSequenceDiagnostics` will follow soon

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44316

cmsbuild commented:
Pull request #47171 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44326

cmsbuild commented:
Pull request #47171 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47171/44327

cmsbuild commented:
Pull request #47171 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

perrotta commented:
@JanChyczynski are you done with the updates now? Can we re-launch the tests?

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-73d852/45317/summary.html
**COMMIT**: 4b8c72c254af6542422adbe13fabc268c3bb5f65
**CMSSW**: CMSSW_15_1_X_2025-04-01-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47171/45317/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 217 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-01-1100+73d852/67784/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-01-1100+73d852/67784/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3913985
 - DQMHistoTests: [Total failures: 53](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-01-1100+73d852/67784/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3913912
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-01-1100+73d852/67784/triggerResults): no differences found

perrotta commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @rappoccio, @mandrenguyen, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
